### PR TITLE
Add echo server with 3 basic implementations

### PR DESCRIPTION
**Pull request by davidfowl**
Getting something working so I can quickly test some ideas.

cc @halter73 @benaadams @tmds

### CODE REVIEW

File: src/BenchmarksApps/TcpEcho/Program.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                socket.Bind(endpoint);
+                socket.Listen(backlog);
+
+                using var reg = cancellationToken.Register(() => socket.Dispose());
+
+                try
+                {
+                    while (true)
+                    {
+                        var connection = await socket.AcceptAsync();
```
Comment (halter73) [inline src/BenchmarksApps/TcpEcho/Program.cs:68]: How are you driving load?

File: src/BenchmarksApps/TcpEcho/SocketConnection.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            var buffer = new byte[4096];
+            while (true)
+            {
+                var read = await _socket.ReceiveAsync(buffer, SocketFlags.None);
+                if (read == 0)
+                {
+                    break;
+                }
+
+                await _socket.SendAsync(buffer, SocketFlags.None);
```
Comment (tmds) [inline src/BenchmarksApps/TcpEcho/SocketConnection.cs:37]: Slice buffer to `read`?

File: src/BenchmarksApps/TcpEcho/SocketDuplexPipeConnection.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                {
+                    var buffer = input.Writer.GetMemory(2048);
+                    var read = await socket.ReceiveAsync(buffer, SocketFlags.None);
+
+                    if (read == 0)
+                    {
+                        break;
+                    }
+
+                    await input.Writer.FlushAsync();
```
Comment (tmds) [inline src/BenchmarksApps/TcpEcho/SocketDuplexPipeConnection.cs:41]: same here, use `read`?

File: src/BenchmarksApps/TcpEcho/SocketPipeConnection.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                {
+                    var buffer = pipe.Writer.GetMemory(2048);
+                    var read = await _socket.ReceiveAsync(buffer, SocketFlags.None);
+
+                    if (read == 0)
+                    {
+                        break;
+                    }
+
+                    await pipe.Writer.FlushAsync();
```
Comment (tmds) [inline src/BenchmarksApps/TcpEcho/SocketPipeConnection.cs:41]: and here ...

File: src/BenchmarksApps/TcpEcho/SocketDuplexPipeConnection.cs (hunk 3 above)
Comment (halter73) [reply src/BenchmarksApps/TcpEcho/SocketDuplexPipeConnection.cs:41]: Good catch. Given how prescriptive the default Pipe implementation can be fore some things, it's surprising that it's OK with a missing call to PipeWriter.Advance(). I'm guessing nothing is actually being written to the socket.

File: src/BenchmarksApps/TcpEcho/Program.cs (hunk 1 above)
Comment (davidfowl) [reply src/BenchmarksApps/TcpEcho/Program.cs:68]: The intent is to use this https://github.com/aspnet/Benchmarks/pull/1485.

File: src/BenchmarksApps/TcpEcho/SocketConnection.cs (hunk 2 above)
Comment (davidfowl) [reply src/BenchmarksApps/TcpEcho/SocketConnection.cs:37]: oops 😄

File: src/BenchmarksApps/TcpEcho/SocketDuplexPipeConnection.cs (hunk 3 above)
Comment (davidfowl) [reply src/BenchmarksApps/TcpEcho/SocketDuplexPipeConnection.cs:41]: OOps

**Pull request merged**